### ISSUE TITLE
stop podman.service: inspecting object: no such object: "top_etc"

### ISSUE THREAD

**Issue by edsantiago** [containers/podman#17904]
Happens very, very often when using sqlite:
```
  stop podman.service
...
Execing [...]/podman --root .../server_root --runroot .../runroot inspect --format={{.State.Running}} top_WthZBtzo

Error: inspecting object: no such object: "top_WthZBtzo"
```
Important note: this is an "Execing" invocation of podman, completely different from the normal e2e test "Running" invocation. It does not have the usual e2e `--this --that` options. sqlite is selected via containers.conf in this setup.

**Comment by edsantiago** [comment]
From the last few days

* debian-12 : int podman debian-12 root host sqlite
  * PR #17831
    * [03-24 11:25](https://api.cirrus-ci.com/v1/artifact/task/5496385294827520/html/int-podman-debian-12-root-host-sqlite.log.html#t--stop-podman-service--1) in Systemd activate [It] stop podman.service
    * [03-24 11:25](https://api.cirrus-ci.com/v1/artifact/task/5496385294827520/html/int-podman-debian-12-root-host-sqlite.log.html#t--stop-podman-service--2) in Systemd activate [It] stop podman.service
    * [03-24 11:25](https://api.cirrus-ci.com/v1/artifact/task/5496385294827520/html/int-podman-debian-12-root-host-sqlite.log.html#t--stop-podman-service--3) in Systemd activate [It] stop podman.service
* debian-12 : int podman debian-12 rootless host sqlite
  * PR #17831
    * [03-24 19:39](https://api.cirrus-ci.com/v1/artifact/task/6346505854910464/html/int-podman-debian-12-rootless-host-sqlite.log.html#t--stop-podman-service--1) in Systemd activate [It] stop podman.service
* fedora-36 : int podman fedora-36 rootless host sqlite
  * PR #17831
    * [03-23 18:43](https://api.cirrus-ci.com/v1/artifact/task/6116601657819136/html/int-podman-fedora-36-rootless-host-sqlite.log.html#t--stop-podman-service--1) in Systemd activate [It] stop podman.service
    * [03-23 18:43](https://api.cirrus-ci.com/v1/artifact/task/6116601657819136/html/int-podman-fedora-36-rootless-host-sqlite.log.html#t--stop-podman-service--2) in Systemd activate [It] stop podman.service
    * [03-23 18:43](https://api.cirrus-ci.com/v1/artifact/task/6116601657819136/html/int-podman-fedora-36-rootless-host-sqlite.log.html#t--stop-podman-service--3) in Systemd activate [It] stop podman.service
* fedora-37 : int podman fedora-37 root container sqlite
  * PR #17831
    * [03-24 16:57](https://api.cirrus-ci.com/v1/artifact/task/6578743729192960/html/int-podman-fedora-37-root-container-sqlite.log.html#t--stop-podman-service--1) in Systemd activate [It] stop podman.service
* fedora-37 : int podman fedora-37 rootless host sqlite
  * PR #17831
    * [03-25 08:16](https://api.cirrus-ci.com/v1/artifact/task/4557084134473728/html/int-podman-fedora-37-rootless-host-sqlite.log.html#t--stop-podman-service--1) in Systemd activate [It] stop podman.service
    * [03-24 22:38](https://api.cirrus-ci.com/v1/artifact/task/4567033191333888/html/int-podman-fedora-37-rootless-host-sqlite.log.html#t--stop-podman-service--1) in Systemd activate [It] stop podman.service
    * [03-23 16:35](https://api.cirrus-ci.com/v1/artifact/task/6319652947623936/html/int-podman-fedora-37-rootless-host-sqlite.log.html#t--stop-podman-service--1) in Systemd activate [It] stop podman.service
    * [03-23 16:35](https://api.cirrus-ci.com/v1/artifact/task/6319652947623936/html/int-podman-fedora-37-rootless-host-sqlite.log.html#t--stop-podman-service--2) in Systemd activate [It] stop podman.service
    * [03-23 16:35](https://api.cirrus-ci.com/v1/artifact/task/6319652947623936/html/int-podman-fedora-37-rootless-host-sqlite.log.html#t--stop-podman-service--3) in Systemd activate [It] stop podman.service
    * [03-22 20:13](https://api.cirrus-ci.com/v1/artifact/task/5072482357477376/html/int-podman-fedora-37-rootless-host-sqlite.log.html#t--stop-podman-service--1) in Systemd activate [It] stop podman.service
    * [03-22 20:13](https://api.cirrus-ci.com/v1/artifact/task/5072482357477376/html/int-podman-fedora-37-rootless-host-sqlite.log.html#t--stop-podman-service--2) in Systemd activate [It] stop podman.service
    * [03-22 20:13](https://api.cirrus-ci.com/v1/artifact/task/5072482357477376/html/int-podman-fedora-37-rootless-host-sqlite.log.html#t--stop-podman-service--3) in Systemd activate [It] stop podman.service

**Comment by edsantiago** [comment]
Most recent: last night, [f36 root](https://api.cirrus-ci.com/v1/artifact/task/5193331160907776/html/int-podman-fedora-36-root-host-sqlite.log.html#t--stop-podman-service--1), in a CI run that included my ignore-enoent patch.

**Comment by vrothberg** [comment]
OK, I _think_ we have a really bad problem in the e2e tests which may have been hidden by the 3 restarts.

Setting environment variables impacts _all_ goroutines.  So doing an `os.Setenv()` impacts everything.  With a bad timing, tests can shoot each other.

Instead, we need to pass environment variables via `exec.Command`.

@Luap99 can you double check my theory?  I am still a bit slow these days.

**Comment by Luap99** [comment]
I am not 100% on this but I do not think this is a problem, if it were would would see much more weird things.

Reading https://onsi.github.io/ginkgo/#spec-cleanup-aftereach-and-defercleanup shows a pattern similar to ours.
Environment variables are per process and as I see it locally when you run things in parallel (`-nodes 3`) it will actually run them as separate processes:
```
pholzing   41096  1.4  0.1 1854584 36736 pts/1   Sl+  16:16   0:00 /home/pholzing/go/src/github.com/containers/podman/test/e2e/e2e.test --test.timeout=1h30m0s --test.coverprofile=e2e.coverprofile.3 --ginkgo.seed=1680099351 --ginkgo.flakeAttempts=3 --ginkgo.progress --ginkgo.debug --ginkgo.parallel.node=3 --ginkgo.parallel.total=3 --ginkgo.parallel.streamhost=http://127.0.0.1:42835 --ginkgo.parallel.synchost=http://127.0.0.1:42835 --ginkgo.noColor --ginkgo.slowSpecThreshold=5.00000 --ginkgo.v --ginkgo.trace
pholzing   41097  1.4  0.1 1780340 36416 pts/1   Sl+  16:16   0:00 /home/pholzing/go/src/github.com/containers/podman/test/e2e/e2e.test --test.timeout=1h30m0s --test.coverprofile=e2e.coverprofile.1 --ginkgo.seed=1680099351 --ginkgo.flakeAttempts=3 --ginkgo.progress --ginkgo.debug --ginkgo.parallel.node=1 --ginkgo.parallel.total=3 --ginkgo.parallel.streamhost=http://127.0.0.1:42835 --ginkgo.parallel.synchost=http://127.0.0.1:42835 --ginkgo.noColor --ginkgo.slowSpecThreshold=5.00000 --ginkgo.v --ginkgo.trace
pholzing   41099  1.3  0.1 1780596 36580 pts/1   Sl+  16:16   0:00 /home/pholzing/go/src/github.com/containers/podman/test/e2e/e2e.test --test.timeout=1h30m0s --test.coverprofile=e2e.coverprofile.2 --ginkgo.seed=1680099351 --ginkgo.flakeAttempts=3 --ginkgo.progress --ginkgo.debug --ginkgo.parallel.node=2 --ginkgo.parallel.total=3 --ginkgo.parallel.streamhost=http://127.0.0.1:42835 --ginkgo.parallel.synchost=http://127.0.0.1:42835 --ginkgo.noColor --ginkgo.slowSpecThreshold=5.00000 --ginkgo.v --ginkgo.trace
```

Therefore each node has it own set of envs and they should not collide. However if a test forgets Unsetenv() then yes we are leaking the env into following tests.

**Comment by edsantiago** [comment]
Would it help to instrument ginkgo such that it dumps ENV on failure? Noisy, but if someone points me at a starting point I'll add it to my test PR.

**Comment by edsantiago** [comment]
Here we go - this is why we can't enable sqlite system tests.

This is a [failure in my enable-sqlite PR](https://api.cirrus-ci.com/v1/artifact/task/6368320933330944/html/int-podman-fedora-37-root-host-sqlite.log.html). It's an **int** failure, not sys, and it's a triple-ginkgo failure, and this is what I see when sqlite is enabled globally via `containers.conf`.

As I mentioned in comment 0, but this is important and needs repeating, **the podman that runs in this test is not the usual e2e podman**. It does not have the long boring series of `--this-and-that`. Hence, in normal `int` tests, it's actually running boltdb. Now, in #17954, it will run sqlite, and will consistently triple-fail. The podman Exec has `--root` but no `--runroot`, so I thought maybe I can fix it by adding `--runroot`, but nope, still fails. That's as far as I got.

Anyhow. #17954 can't merge until this is fixed.

**Comment by vrothberg** [comment]
That is great news actually.  I couldn't put the pieces together when thinking this was a flake.  A consistent failure makes more sense.

**Comment by edsantiago** [comment]
It _is_ a flake. If it were consistent, none of the `int sqlite` tests would pass. But last night before bed I force-pushed onto #17954, no change whatsoever, and all tests passed on the first try.

**Comment by vrothberg** [comment]
What the ...

I didn´t take a look yet but _why_ would it flake with containers.conf being set?  That is something static.  Given the tests don´t have an unexpected interference (see my theory on CONTAINERS_CONF), then it may be a bug/race in libpod?

**Comment by vrothberg** [comment]
Apologies, I feel almost like abusing you trying out different ideas, @edsantiago.

To test whether tests interfere, could we try running with `GINKGONODES=1` such that tests run in sequence and not in parallel?

**Comment by vrothberg** [comment]
I am more or less thinking out loud at the moment and hope for our collective swarm brain to find a solution.  One thing that strikes me:

```
         Execing /var/tmp/go/src/github.com/containers/podman/bin/podman --root /tmp/podman_test1092595828/server_root --runroot /tmp/podman_test1092595828/runroot inspect --format={{.State.Running}} top_WPsMuMFf
         
         Error: inspecting object: no such object: "top_WPsMuMFf"
         Execing /var/tmp/go/src/github.com/containers/podman/bin/podman --root /tmp/podman_test1092595828/server_root --runroot /tmp/podman_test1092595828/runroot rm -f top_WPsMuMFf
         
         top_WPsMuMFf
```

While the `podman inspect $CTR` fails.  The defered `podman rm $CTR` just works fine (see https://github.com/containers/podman/blob/a7d0408761e0649dad419bd7dd447f68a0662cd4/test/e2e/systemd_activate_test.go#L129).

@edsantiago, can you confirm that it's always the `podman inspect` that fails?  That would somehow rather strongly point to a race inside of Podman rather than test interference.

**Comment by edsantiago** [comment]
It would take me too long to confirm-confirm, I would need to reexamine each log individually... but I _have_ examined each log, I do so day in and day out for my flake categorizing, and I think I would've noticed if there were any variation in the pattern. I'm 80% confident that they all look the same.

**Comment by vrothberg** [comment]
Did it "work"?

**Comment by edsantiago** [comment]
CI still running. One flake, but not the desired one.

**Comment by edsantiago** [comment]
It flaked! And I don't think you're going to like the debug trace but [here ya go](https://api.cirrus-ci.com/v1/artifact/task/5830718668931072/html/int-podman-fedora-36-root-host-sqlite.log.html#t--stop-podman-service--1). I Can only conclude gremlins.

**Comment by vrothberg** [comment]
```
         Execing /var/tmp/go/src/github.com/containers/podman/bin/podman-remote --url tcp://127.0.0.1:34745 inspect --format={{.State.Running}} top_FFxihwus
         
         true
         Execing /var/tmp/go/src/github.com/containers/podman/bin/podman --root /tmp/podman_test3019531123/server_root ps --all
         
         CONTAINER ID  IMAGE       COMMAND     CREATED     STATUS      PORTS       NAMES
         Execing /var/tmp/go/src/github.com/containers/podman/bin/podman --root /tmp/podman_test3019531123/server_root inspect --format={{.State.Running}} top_FFxihwus
         
         Error: inspecting object: no such object: "top_FFxihwus"
         Execing /var/tmp/go/src/github.com/containers/podman/bin/podman --root /tmp/podman_test3019531123/server_root rm -f top_FFxihwus
         
         top_FFxihwus
```

OK:
* `podam-remote inspect` sees the container
* `podman ps --all` doesn't see it (yet?)
* `podman inspect` doesn't see it (yet?)
* the deferred `podman rm` sees it

That smells like a consistency issue in sqlite where when running in `podman system service` the container is NOT written to the file immediately but at some point.

@mheon PTAL

**Comment by edsantiago** [comment]
Yes, see comments above about #17954  and triple-failures and blocking.

**Comment by edsantiago** [comment]
I can't reproduce. I tried on my laptop and on 1mt. Tried infinite-looping the `e2e` test. Tried a lot of different approaches at distilling the relevant aspects of the code. Nothing. 

It's either something specific to the CI VMs, or something specific to e2e test setup. Or something else. I'll have to get back to this next week.

**Comment by vrothberg** [comment]
Thanks, Ed!  I am also checking it at the moment and couldn't spot any difference wrt to TMP environments.

**Comment by edsantiago** [comment]
This is weird. Usual failure [in my no-flake-repeat PR](https://api.cirrus-ci.com/v1/artifact/task/4961294009237504/html/int-podman-fedora-36-rootless-host-sqlite.log.html#t--stop-podman-service--1) except this time there's a spurious error message right in the middle of the log:
```
Error: unable to connect to Podman socket: Get "http://d/v4.5.0/libpod/_ping": ssh: rejected: connect failed (open failed)
                                                                          -----^^^
```
**ssh?!?**

This is my hammer-sqlite PR, run with `ginkgo -nodes 1`, so it's (presumably) not a concurrent job somewhere else.

It doesn't cause a test failure (even though it says "Error", there's no actual ginkgo error associated with it).

My working assumption is that this is a leftover background job from a previous test. If so, I fail to understand how it could affect this test: presumably it uses a different `--root` so it can't (?) affect this test's sqlite db. So I'm tentatively dismissing it as a red herring but am posting anyway in case someone else wants to look.

I'm going to repeat once again that this is a test that does not use the `--this --that` podman invocation, and I have a weird obsession with the thought that this is not coincidence. Something about `podman --this --that` (the usual e2e invocation) prevents this flake. I'm quite possibly wrong, but until I see this flake in a  `podman -this -that` invocation I'm going to keep insisting in hopes that someone will pay attention.

**Comment by vrothberg** [comment]
> I'm going to repeat once again that this is a test that does not use the --this --that podman invocation, and I have a weird obsession with the thought that this is not coincidence. Something about podman --this --that (the usual e2e invocation) prevents this flake. I'm quite possibly wrong, but until I see this flake in a podman -this -that invocation I'm going to keep insisting in hopes that someone will pay attention.

Have you seen this test flake with `boltdb`?

While we cannot exclude the usual `--this --that` to be causing the flake, it's not on the top of my suspect list.

The symptom we see is:
* `podman-remote` create, starts and inspect container `C`
* Right after the local non-remote `podman inspect` does not see container `C`
  * But it's flaking, so we see it sometimes
  * The deferred `podman rm` then sees `C` some time after the failed `inspect`

To me, it screams like `sqlite` doesn't write the changes immediately to disk after a transaction.

> I'm quite possibly wrong, but until I see this flake in a podman -this -that invocation I'm going to keep insisting in hopes that someone will pay attention.

I hear/read you :)   `--root` should be enough for separating tests.

@Luap99 WDYT?

**Comment by edsantiago** [comment]
>Have you seen this test flake with boltdb?

No, this is brand-new in sqlite, and *only* when sqlite is selected via `containers.conf` (i.e. it only happens in the nontraditional podman invocation, the `Exec` one, never in the `--this --that` one #deadhorse)

**Comment by vrothberg** [comment]
> > Have you seen this test flake with boltdb?
> 
> No, this is brand-new in sqlite, and _only_ when sqlite is selected via `containers.conf` (i.e. it only happens in the nontraditional podman invocation, the `Exec` one, never in the `--this --that` one #deadhorse)

Thanks for confirming.

To elaborate on my theory:  this is the first time we have a mixed `podman-remote` and `podman` test with `sqlite`.  Only with the `containers.conf` change this test was run with `sqlite` since the `--this --that` is missing here.  I'd expect the `apiv2` tests to also bark with `sqlite` as there's a mixed use of `curl` (i.e., talking to a long-running `podman system service`) and non-remote `podman`.

**Comment by edsantiago** [comment]
Except that the flow is actually:
1. start podman-remote server; create and start container using podman-remote
2. **kill the server**
3. run local-podman inspect

https://github.com/containers/podman/blob/e1790793517aa65b729896749384c9dee5a4622e/test/e2e/systemd_activate_test.go#L133-L138

I could juuuuust barely accept that a running process could delay in flushing the db. I'm having a harder time accepting that from a terminated process. But maybe I'm misreading the code.

**Comment by Luap99** [comment]
Lets patch the `--this --that` bug and then we can take a look again if it is actually sqlite or just something weird where this is a db missmatch.

And arguably we should prevent users from mismatching the db. IMO podman should refuse to run if we detect a botdb but sqlite is selected and the other way around but I guess this is a discussion for a another issue.

**Comment by edsantiago** [comment]
>Lets patch the --this --that bug

It's not entirely fixable, because of the `socket-activate` invocation.

More importantly, I'm ambivalent about "fixing" it: I think that might end up hiding a real problem. But I can't back that with any evidence, because this flake is just so baffling.

FWIW here is a typical invocation showing all the options used in a ginkgo podman call (reordered and reformatted for legibility):
```
podman --storage-opt vfs.imagestore=/tmp/imagecachedir
       --root    /tmp/podman_test3884874678/root
       --runroot /tmp/podman_test3884874678/runroot
       --tmpdir  /tmp/podman_test3884874678
       --runtime crun
       --conmon /usr/bin/conmon
       --network-config-dir /home/some26440dude/.config/cni/net.d
       --network-backend cni
       --cgroup-manager systemd
       --events-backend file
       --db-backend sqlite
       --storage-driver vfs
```
Is it possible that `--tmpdir` is propagating into the sqlite code, and without it there's (waves hands) some sort of conflict where multiple concurrent podman processes, using the same default TMPDIR, clobber each other? #magicalthinking

**Comment by vrothberg** [comment]
That's quite a big list.

I think we can refactor creating all podman arguments into a function that can be used by the ordinary e2e tests and this specific one.

> Is it possible that --tmpdir is propagating into the sqlite code, and without it there's (waves hands) some sort of conflict where multiple concurrent podman processes, using the same default TMPDIR, clobber each other? #magicalthinking

`TMPDIR` impacts sqlite but I am unsure about `--tmpdir`.  I checked the processes last week and their environment didn't differ wrt. TMPDIR.

What's really curious to me is that `podman inspect` at time t0 does not work (sometimes) but `podman rm` at time tN does work and see the container.

**Comment by edsantiago** [comment]
Data point: it's not `--tmpdir`. I added that to my PR, and [flake](https://api.cirrus-ci.com/v1/artifact/task/6395980858261504/html/int-podman-fedora-36-rootless-host-sqlite.log.html#t--stop-podman-service--1), [flake](https://api.cirrus-ci.com/v1/artifact/task/5270080951418880/html/int-podman-fedora-37-rootless-host-sqlite.log.html#t--stop-podman-service--1) anyway.

**Comment by vrothberg** [comment]
I just sat down to massage the test to use `--this --that`

**Comment by vrothberg** [comment]
> I just sat down to massage the test to use `--this --that`

PR https://github.com/containers/podman/pull/18056

**Comment by edsantiago** [comment]
Well, I screwed up. I should've looked up the ginkgo description of [Eventually()](https://onsi.github.io/gomega/#eventually). TL;DR it's async, so the socket server is very likely still running concurrently with the podman-local process. [Proof](https://api.cirrus-ci.com/v1/artifact/task/6332802426208256/html/int-podman-fedora-36-rootless-host-sqlite.log.html#t--stop-podman-service--1):
```
Execing [server]
Execing [podman-remote create & start]
Execing [podman-local inspect]
...
* Failure, timed out, Expected [systemd-socket] process to exit. It did not.
```
Anyhow, I see #18056 and will take a look at it now.

**Comment by vrothberg** [comment]
I am not sure, @edsantiago.  Here it fails because the `system service` could not be killed with the given timeout.  Doesn't that mean that in other cases, the service has been killed successfully?

**Comment by edsantiago** [comment]
It means that in other cases the service *Eventually()* gets killed successfully. But we don't know when. It's possible (and IMHO likely) for this to happen:
1. signal
2. podman-local inspect (fails)
3. socket process exits, flushing DB
4. podman-local rm (succeeds)

**Comment by vrothberg** [comment]
Ah, OK.  Yes, that would still be in line with the theory of the DB not writing the changes to disk immediately.

BUT the `podman-local rm` seems to always succeed - even then the `local inspect` fails.

**Comment by edsantiago** [comment]
`s/Eventually/Expect/` in my PR #17831, and I now expect it to pass.

(I'm not saying that's what we want to do! If sqlite is not flushing writes, that's a Very Very Bad Thing that we need to pursue! I'm just saying I want to keep gathering data)

**Comment by vrothberg** [comment]
Agreed.  I want to get to the bottom of it.  It's just too suspicious that it does not fail with bolt.

**Comment by edsantiago** [comment]
It took about five retries, but even with `Eventually` changed to `Expect`, it flaked. [Debian](https://api.cirrus-ci.com/v1/artifact/task/6016272496852992/html/int-podman-debian-12-rootless-host-sqlite.log.html#t--stop-podman-service--1) and [F37](https://api.cirrus-ci.com/v1/artifact/task/5805166264320000/html/int-podman-fedora-37-root-host-sqlite.log.html#t--stop-podman-service--1). Giving up for the day.

**Comment by edsantiago** [comment]
New variant (I think): seeing in the `rm`. [f36 root](https://api.cirrus-ci.com/v1/artifact/task/6721280674627584/html/int-podman-fedora-36-root-host-sqlite.log.html#t--stop-podman-service--1):
```
Execing podman [...] inspect --format={{.State.Running}} top_gAXKGPfR
Error: inspecting object: no such object: "top_gAXKGPfR"
Execing podman [...] rm -f top_gAXKGPfR
Error: removing storage for container "top_gAXKGPfR": 1 error occurred:
	* unlinkat /tmp/podman_test3195926079/server_root/overlay-containers/CID/userdata/shm: device or resource busy
```

**Comment by vrothberg** [comment]
I think that's the EBUSY flake.  I've seen it a while ago in this PR (https://github.com/containers/podman/issues/17904#issuecomment-1490206809).

**Comment by edsantiago** [comment]
Yes, it's unlinkat/EBUSY ... but this is happening in `Execing podman`, the one without `--this --that`. It's a brand-new failure mode, because this is happening in a simple clean setup: custom rootdir, very few distracting command-line options. (Which of course is the case in system tests, where we mostly see unlinkat/EBUSY, but usually in system tests we see it near the end of tests, and of course system tests use the default `/var/containers/whatever`, so there's always been the possibility that some previous test screws up state for subsequent ones).

Anyhow, this does not give me any useful information, it's just a hmmmm.

**Comment by edsantiago** [comment]
Still [seeing it](https://api.cirrus-ci.com/v1/artifact/task/4964221834756096/html/int-podman-fedora-37-rootless-host-sqlite.log.html#t--stop-podman-service--1) on main which currently includes #18083 as well as `sync=EXTRA`.

**Comment by vrothberg** [comment]
With https://github.com/containers/podman/pull/18056 merged, let's close this issue.
